### PR TITLE
Assignment-1 Web API Demo (Books API) 05/09/2019

### PR DESCRIPTION
**Pull request by tavisca-nrajpurohit**
@tavisca-shitaljit Please review :)

 **GET POST PUT DELETE** on three layers - **API, Service and Data level**. 
Used Json File for **persistent storage** of data. 
Used **custom response object**. 
Written **test cases** for **service layer and data layer** functionality.

### CODE REVIEW

File: WebAPI.Tests/WebAPITest.cs [hunk 1]
```diff
@@ -0,0 +1,164 @@
...
+
+            Assert.True(booklist.Count > 0);
+        }
+
+        [Fact]
+        public void Get_With_Parameters_Test_All_Favourable_Inputs()
+        {
+            Response response = bookservice.Get(1);
+            List<Book> booklist = response.booklist;
+
```
Comment (tavisca-shitaljit) [inline WebAPI.Tests/WebAPITest.cs:30]: Instead of count comparison you can compare Id of the returned book

**Review by tavisca-shitaljit — CHANGES_REQUESTED** [review]
Good work.

Let's refactor a bit to see how can we reduce lines of codes and add some more scenarios in test cases.

File: WebAPI.Tests/WebAPITest.cs (hunk 1 above)
Comment (tavisca-nrajpurohit) [reply WebAPI.Tests/WebAPITest.cs:30]: Sure, I'll Update that functionality.

File: WebAPI.Tests/WebAPITest.cs [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+            Assert.Equal("Error: No such Book Found in the Repository", msg);
+        }
+
+        [Fact]
+        public void POST_With_Everything_Good()
+        {
+            Book book = new Book() { id =15,price=140,title="Book twelve",author="Neelesh ji",category="Fiction"};
+            Response response = bookservice.Post(book);
+            List<Book> booklist = response.booklist;
+            string msg = response.message;
```
Comment (tavisca-snarhe) [inline WebAPI.Tests/WebAPITest.cs]: you can name msg to responseMessage

File: WebAPI.Tests/WebAPITest.cs [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+
+            Assert.Equal("Data Added Successfully !", msg);
+            response = bookservice.Delete(15);
+        }
+
+        [Fact]
+        public void POST_Input_with_negetive_Id_and_price()
```
Comment (tavisca-snarhe) [inline WebAPI.Tests/WebAPITest.cs]: can rename Id to bookId

File: WebAPI.Tests/WebAPITest.cs [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+        {
+            Book book = new Book() { id = -1, price = -140, title = "Book twelve", author = "Neelesh ji", category = "Fiction" };
+            Response response = bookservice.Post(book);
+            List<Book> booklist = response.booklist;
+            string msg = response.message;
+            Assert.Equal("Error: Invalid Book ID or Price, they should be a positive number!", msg);
+        }
+
+        [Fact]
+        public void POST_Input_With_NON_ALphabet_Book_Title()
```
Comment (tavisca-snarhe) [inline WebAPI.Tests/WebAPITest.cs]: why is NON_AL in caps?

File: WebApiDemo/Model/BookService.cs [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+                    return GenerateResponse(null, "Error: Invalid Book Title, Author or Category. Please use only Alphabets for them!");
+                }
+            }
+            else
+            {
+                return GenerateResponse(null, "Error: Invalid Book ID or Price, they should be a positive number!");
+            }
+        }
+
+        private bool isAlphabets(string str)
```
Comment (tavisca-snarhe) [inline WebApiDemo/Model/BookService.cs:70]: using regex here will helpfull.

File: WebApiDemo/Model/BookService.cs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+            foreach (char c in str)
+            {
+                if (!Char.IsLetter(c) && c!=' ')
+                    return false;
+            }
+            return true;
+        }
+
+        public Response Put(int id, Book book)
+        {
```
Comment (tavisca-snarhe) [inline WebApiDemo/Model/BookService.cs:81]: construct errors and lastly generate the response with all the errors.

**Review by tavisca-snarhe — CHANGES_REQUESTED** [review]

File: WebAPI.Tests/WebAPITest.cs (hunk 2 above)
Comment (tavisca-nrajpurohit) [reply WebAPI.Tests/WebAPITest.cs]: sure

File: WebAPI.Tests/WebAPITest.cs (hunk 3 above)
Comment (tavisca-nrajpurohit) [reply WebAPI.Tests/WebAPITest.cs]: sure

File: WebAPI.Tests/WebAPITest.cs (hunk 4 above)
Comment (tavisca-nrajpurohit) [reply WebAPI.Tests/WebAPITest.cs]: thanks for pointing this out.

File: WebApiDemo/Model/BookService.cs (hunk 5 above)
Comment (tavisca-nrajpurohit) [reply WebApiDemo/Model/BookService.cs:70]: sure

**Comment by tavisca-nrajpurohit** [comment]
Updated code with necessary changes. and also updated functionality for  returning relevant status codes from the API

**Pull request closed without merging**